sklearn/inspection/_partial_dependence.py

```diff
@@ -679,7 +687,8 @@ def convert_feature(fx):
     deciles = {}
     for fx in chain.from_iterable(features):
         if fx not in deciles:
-            deciles[fx] = mquantiles(X[:, fx], prob=np.arange(0.1, 1.0, 0.1))
+            X_col = _safe_indexing(X, fx, axis=1)
```

Comment:
Why do we need this? It doesn't seem right that we need to change the indexing just because we are automatically inferring the names now

Reply:
`X[:, fx]` should be `X[fx]` for a dataframe. `safe_indexing` takes care about calling the right method.

Reply:
I don't see where you are changing anything to the `features` variable so I don't understand why you need this. 

`fx` is still an int in this PR.

Reply:
Yes fx is an integer so the following will fail with dataframe:


```python

In [1]: import pandas as pd                                                          

In [2]: import numpy as np                                                           

In [3]: X = np.random.randn(10, 10)                                                  

In [4]: df = pd.DataFrame(X)                                                         

In [5]: df[:, 0]                                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-c4a594bb2bd7> in <module>
----> 1 df[:, 0]

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/frame.py in __getitem__(self, key)
   2993             if self.columns.nlevels > 1:
   2994                 return self._getitem_multilevel(key)
-> 2995             indexer = self.columns.get_loc(key)
   2996             if is_integer(indexer):
   2997                 indexer = [indexer]

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/range.py in get_loc(self, key, method, tolerance)
    377             except ValueError:
    378                 raise KeyError(key)
--> 379         return super().get_loc(key, method=method, tolerance=tolerance)
    380 
    381     @Appender(_index_shared_docs["get_indexer"])

~/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2895                 )
   2896             try:
-> 2897                 return self._engine.get_loc(key)
   2898             except KeyError:
   2899                 return self._engine.get_loc(self._maybe_cast_indexer(key))

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

TypeError: '(slice(None, None, None), 0)' is an invalid key
```

Reply:
I thought https://github.com/scikit-learn/scikit-learn/pull/14028 was adding support for dataframes

Reply:
> I thought #14028 was adding support for dataframes

You are right and it works by specifying `features` and `feature_names`. `plot_partial_dependence` was simply converting the dataframe into a NumPy array and storing the columns info with `features` and `feature_names`.

The idea now is to be able to infer columns name when giving a dataframe (and avoid any data copy if possible). So we can avoid the `check_array` to convert the dataframe into a numpy array and then use the columns name. However, this line was intended to be used only with numpy aray while now a dataframe or a numpy array can get there.

Reply:
I was missing the conversion above, thanks